### ISSUE TITLE
Tackling Multi-Input Layers in Pipeline Parallelism and Eliminating Redundant Communication

### ISSUE THREAD

**Issue by eee4017** [PaddlePaddle/Paddle#60861]
### 需求描述 Feature Description

# Introduction
The current `PipelineLayer` design inadvertently leads to a lack of awareness among programmers about the transmission of certain data across GPUs. This oversight can result in inefficiencies, particularly in large-scale models where inter-GPU communication is a significant bottleneck.

# Background
When implementing pipeline parallelism in deep learning models developers typically utilize a PipelineLayer to distribute various layers across multiple GPUs. This mechanism handles the input and output transmission between GPUs. Consider the following basic structure:

```python
class PipelineLayer:
    ...
    def forward(x):
       ...
       for layer in layers:
           x = layer(x)

model = PipelineLayer(layers=[Layer1, Layer2_1, Layer2_2, Layer2_3])
output = model(x)
```

In this setup, each layer, such as Layer1, Layer2_1, etc., receives a single input (e.g., Layer1(x)). However, a limitation arises with PipelineLayer as it currently supports only layers that require a single input. This constraint becomes problematic when the model architecture necessitates layers that demand multiple inputs.

To address this challenge, we propose a workaround involving the packaging of inputs into a tuple of tensors. We then introduce an UnpackLayer, which unpacks these inputs and feeds them into the respective layer. Here's an illustration of this approach:

```python
class UnpackLayer:
    ...
    def forward(args):
        x, mask = args
        return layer(x, mask)

model = PipelineLayer(layers=[
    UnpackLayer(Layer1),
    UnpackLayer(Layer2_1),
    UnpackLayer(Layer2_2),
    UnpackLayer(Layer2_3)
])

input = (x, mask)
output = model(input)

```
An implementation example of `UnpackLayer` can be found in the PaddleNLP framework ([GPTDecoderLayerPipe](https://github.com/PaddlePaddle/PaddleNLP/blob/develop/paddlenlp/transformers/gpt/modeling_pp.py#L117)), which unpacks hidden_states, attention_mask, and position_ids from a tuple of tensors.

However, a critical observation here is that the PipelineLayer automatically transfers the packed input (e.g., (x, mask) in the above example) between GPUs. This automatic transfer can lead to inefficiencies, especially when certain elements of the input, like the `attention_mask` in casual mode for GPT models, remain constant and do not necessitate inter-GPU communication. By deploying a temporary solution to remove this redundant mask communication, we've noted a significant throughput enhancement, roughly 20%, when training 1.3B GPT models on 4 A100 GPUs using pipeline parallelism.

# Discussion on Possible Solutions

### Model Design

In the [GPT model in PaddleNLP](https://github.com/PaddlePaddle/PaddleNLP/blob/develop/paddlenlp/transformers/gpt/modeling_pp.py), explore methods to share the `attention_mask` among layers, thus reducing unnecessary data transmission and enhancing efficiency.

### Framework Design
1. The modification of PipelineLayer to handle multiple inputs, allowing for a clearer description of tensors and their inter-GPU sharing requirements, thereby increasing developer awareness and reducing unnecessary data transfers. 
2. The introduction of a tensor tagging system to control the transmission of data across GPUs. This enables specific marking of tensors to prevent redundant communication between GPUs.

**Comment by ForFishes** [comment]
Regarding PipelineLayer, it inherently supports the passing of tuple tensors, similar to the functionality of the UnpackLayer that you mentioned. The PaddleNLP framework is further encapsulated on this basis.

As for the issue of redundant communication, there isn't currently an effective solution. The reason is as follows:
For dynamic graphs, there is no full graph information prior to execution. Moreover, when constructing the model, there is no information available about the execution phase (such as the micro-batch id), which means that this problem cannot be resolved. Addressing it during the execution phase would require users to write cross-gpu communication methods and directly transfer data, which also necessitates knowledge of the state during execution.
Given this, we haven't yet found a universal, user-friendly approach to help users achieve this efficiently. While a customized approach to optimize a specific aspect is feasible by rewriting the pipeline scheduling, it's not a universal solution.
